agent/hcp/deps.go

```diff
@@ -23,5 +32,80 @@ func NewDeps(cfg config.CloudConfig, logger hclog.Logger) (d Deps, err error) {
 	}
 
 	d.Provider, err = scada.New(cfg, logger.Named("hcp.scada"))
+	if err != nil {
+		return
+	}
+
+	d.Sink = sink(d.Client, &cfg, logger)
+
 	return
 }
+
+// sink provides initializes an OTELSink which forwards Consul metrics to HCP.
+// The sink is only initialized if the server is registered with the management plane (CCM).
+// This step should not block server initialization, so errors are logged, but not returned.
+func sink(hcpClient hcpclient.Client, cfg hcpclient.CloudConfig, logger hclog.Logger) *telemetry.OTELSink {
+	ctx := context.Background()
+	ctx = hclog.WithContext(ctx, logger)
+	u, err := verifyCCMRegistration(ctx, hcpClient)
+	if err != nil {
+		logger.Error("failed to verify CCM registration: %w", err)
+		return nil
+	}
+
+	// if endpoint is empty, no metrics endpoint configuration for this Consul server
+	// (e.g. not registered with CCM or feature flag to control rollout) so do not enable the HCP metrics sink.
+	if u == nil {
+		return nil
+	}
+
+	metricsClient, err := hcpclient.NewMetricsClient(cfg, ctx)
+	if err != nil {
+		logger.Error("failed to init metrics client: %w", err)
+		return nil
+	}
+
+	sinkOpts := &telemetry.OTELSinkOpts{
+		Ctx:    ctx,
+		Reader: telemetry.NewOTELReader(metricsClient, u, 10*time.Second),
+	}
+
+	sink, err := telemetry.NewOTELSink(sinkOpts)
+	if err != nil {
+		logger.Error("failed to init OTEL sink: %w", err)
+		return nil
+	}
+
+	return sink
+}
+
+// verifyCCMRegistration checks that a server is registered with the HCP management plane
+// by making a HTTP request to the HCP TelemetryConfig endpoint.
+// If registered, it returns the endpoint for the HCP Telemetry Gateway endpoint where metrics should be forwarded.
+func verifyCCMRegistration(ctx context.Context, client hcpclient.Client) (*url.URL, error) {
```

Comment:
Lets not mention CCM as part of the function name here. I see this function doing two things: 
  1. checking if we can retrieve a valid telemetry config returning 
  2. the url for configuring OTLP metrics exporting.

What do you think about moving the logic for these two things into functions on the `TelemetryConfig` struct? I would propose something like this:
```
func (tc *TelemetryConfig) Enabled() bool  {...}
func (tc *TelemetryConfig) MetricsURL() (*url.URL, error)
```

Then we could remove this method (or just refactor it to get and return the `TelemetryConfig`) and the caller which builds the sink could use those methods on the struct directly (and remove the nil url check).
```
if !telemetryCfg.Enabled() { return nil}

u, err := telemetryCfg.MetricsURL()
if err != nil { ... }
```

Reply:
I like that idea! It would definitely clean up lots of the code. 

To decide if telemetry is enabled or not, we'd need to actually verify that the endpoints are not empty, i.e. do this part:

```go
	endpoint := telemetryCfg.Endpoint
	if override := telemetryCfg.MetricsConfig.Endpoint; override != "" {
		endpoint = override
	}
```
To then parse the URL after in a separate function felt like I'd need to repeat the logic, unless I hold the updated endpoint in the `telemetryCfg.Endpoint` field. Instead, I simplified to have a single function on the telemetryCfg struct.

I updated a latest commit with what this looks like, completely removed the `verifyCCMRegistration` function. Let me know what you think!